pkg/streamingpromql/operators/selectors/extend_range_vector.go

```diff
@@ -0,0 +1,210 @@
+// SPDX-License-Identifier: AGPL-3.0-only
+// Provenance-includes-location: https://github.com/prometheus/prometheus/blob/main/promql/engine.go
+// Provenance-includes-license: Apache-2.0
+// Provenance-includes-copyright: The Prometheus Authors
+
+package selectors
+
+import (
+	"github.com/prometheus/prometheus/promql"
+
+	"github.com/grafana/mimir/pkg/streamingpromql/types"
+	"github.com/grafana/mimir/pkg/util/limiter"
+)
+
+type headTailIterator struct {
+	head, tail []promql.FPoint
+	idx        int // combined iterator position across both
+	lazyLen    int // combined length, lazy initialized in len()
+}
+
+func (i *headTailIterator) len() int {
+	// avoid having to do len(i.head) + len(i.tail) on each call
+	if i.lazyLen == 0 && (len(i.head) > 0 || len(i.tail) > 0) {
+		i.lazyLen = len(i.head) + len(i.tail)
+	}
```

Comment:
I wouldn't bother with this, unless we can see it causes a significant performance problem.

Alternatively, if we retain a reference to the view, then we can get the size from there cheaply.